ipld/get_namespaced_shares.go

```diff
@@ -0,0 +1,185 @@
+	}
+	return ok
+}
+
+// getLeavesByNamespace returns all the leaves from the given root with the given namespace.ID.
+// If nothing is found, it returns both data and error as nil.
+// The following implementation is based on `GetShares`.
+func getLeavesByNamespace(
+	ctx context.Context,
+	bGetter blockservice.BlockGetter,
+	root cid.Cid,
+	nID namespace.ID,
+	maxShares int,
+) ([]ipld.Node, error) {
+	err := SanityCheckNID(nID)
+	if err != nil {
+		return nil, err
+	}
+
+	// we don't know where in the tree the leaves in the namespace are,
+	// so we keep track of the bounds to return the correct slice
+	// maxShares acts as a sentinel to know if we find any leaves
+	bounds := fetchedBounds{int64(maxShares), 0}
+
+	// buffer the jobs to avoid blocking, we only need as many
+	// queued as the number of shares in the second-to-last layer
+	jobs := make(chan *job, (maxShares+1)/2)
+	jobs <- &job{id: root}
+
+	// the wg counter cannot be pre-allocated either, it is incremented with each job
+	var wg wrappedWaitGroup
+	wg.Add(1)
+
+	// we overallocate space for leaves since we do not know how many we will find
+	// on the level above, the length of the Row is passed in as maxShares
+	leaves := make([]ipld.Node, maxShares)
+
+	// stillWorking will be set to false when the waitgroup counter reaches 0, closing the jobs channel
+	for stillWorking := true; stillWorking; {
+		select {
+		case j, ok := <-jobs:
+			if !ok {
+				stillWorking = false
+				break
+			}
+			namespacePool.Submit(func() {
+				// if the WaitGroup counter is above 0, we are still walking the tree
+				defer wg.Done(jobs)
+
+				rootH := plugin.NamespacedSha256FromCID(j.id)
+				if nID.Less(nmt.MinNamespace(rootH, nID.Size())) || !nID.LessOrEqual(nmt.MaxNamespace(rootH, nID.Size())) {
+					return
+				}
+
+				nd, err := plugin.GetNode(ctx, bGetter, j.id)
+				if err != nil {
+					return
+				}
+
+				links := nd.Links()
+				linkCount := uint64(len(links))
+				// wg counter needs to be incremented **before** adding new jobs
+				if linkCount > 1 {
+					wg.Add(int64(linkCount))
+				} else if linkCount == 1 {
+					// successfully fetched a leaf belonging to the namespace
+					leaves[j.pos] = nd
+					// we found a leaf, so we update the bounds
+					// the update routine is repeated until the atomic swap is successful
+					for ok := false; !ok; {
+						ok = bounds.Update(int64(j.pos))
+					}
+					return
+				}
+
+				// this node has links in the namespace, so keep walking
+				for i, lnk := range links {
+					select {
+					case jobs <- &job{
+						id: lnk.Cid,
+						// position represents the index in a flattened binary tree,
+						// so we can return a slice of leaves in order
+						pos: j.pos*2 + i,
+					}:
+					case <-ctx.Done():
+						return
+					}
+				}
+			})
+		case <-ctx.Done():
+			return nil, nil
+		}
+	}
+
+	// we didn't find any leaves, return nil
+	if bounds.lowest == int64(maxShares) {
+		return nil, nil
```

Comment:
I think we should have an error in this case. Something similar to https://github.com/celestiaorg/celestia-node/blob/2cc807743a16f4a955f6d9c819d829f98e6fbbed/header/interface.go#L65

Reply:
History note: we had something similar before but decided to get rid of it in favour of making `len(0)` returned slice a success case when @tzdybal chimed in during development of DA API

Reply:
I don't feel strongly about any of the options, but we might reconsider this. @vgonkivs check out higher-level GetShareByNamespace to see how introduction of such error may affect the code

Reply:
I don't think we should return an error. It sounds like a normal case that querying a namespace will have no shares